tests/integration/retries.py

```diff
@@ -0,0 +1,214 @@
+
+    This clock is resilient in the face of system time adjustments, with the caveat that times can only be compared
+    against each other. Comparing times yields the difference between them, measured in seconds.
+    """
+
+    __slots__ = ("_clock",)
+
+    def __init__(self) -> None:
+        self._clock = time.perf_counter
+
+    def time(self) -> float:
+        """Return the current time.
+
+        The value returned can be compared to other times returned by this method. The difference represents the
+        number of seconds (potentially with a fractional component) that elapsed between them.
+
+        :return: A value that represents the current time.
+        """
+        return self._clock()
+
+    def sleep(self, seconds: float) -> None:
+        """Delay execution for a given number of seconds.
+
+        The argument may have a fractional component for sub-second precision.
+
+        This method is blocking; it is not async-aware.
+
+        :param seconds: The duration to wait before returning from this method.
+        """
+        time.sleep(seconds)
+
+
+def _default_timeout_dilation() -> float:
+    """Load the default timeout dilation factor.
+
+    This factor is used to scale timeouts by a default, which can be useful in environments where tests are slower than
+    expected.
+
+    :returns: the default time dilation to use for timeouts.
+    """
+    dilation = float(os.environ["TIMEOUT_DILATION"]) if "TIMEOUT_DILATION" in os.environ else 1.0
+    if dilation != 1.0:
+        logger.info(f"Non-default timeout dilation: all timeouts will be scaled by a factor of {dilation}.")
+    return dilation
+
+
+P = ParamSpec("P")
+T = TypeVar("T")
+
+
+class _Retrier:
+    __slots__ = ("_on", "_is_retryable", "_timeout", "_clock", "_min_attempts", "_max_attempts")
+
+    def __init__(
+        self,
+        *,
+        on: Sequence[type[BaseException]] | None,
+        is_retryable: Callable[[BaseException], str | None] | None,
+        timeout: dt.timedelta | None,
+        clock: Clock,
+        timeout_dilation: float,
+        min_attempts: int,
+        max_attempts: int | None = None,
+    ) -> None:
+        has_allowlist = on is not None
+        has_callback = is_retryable is not None
+        if has_callback == has_allowlist:
+            raise SyntaxError('either on=[Exception] or callback=lambda x: .. is required')
+        if timeout is None and max_attempts is None:
+            raise SyntaxError('either timeout=... or max_attempts=... is required')
+        if max_attempts is not None and max_attempts < min_attempts:
+            msg = f'max_attempts=x must be equal or higher to min_attempts ({min_attempts})'
+            raise SyntaxError(msg)
+
+        self._on = on
+        self._is_retryable = is_retryable
+
+        dilated_timeout = timeout * timeout_dilation if timeout is not None else None
+        self._timeout = dilated_timeout
+        self._clock = clock
+        self._min_attempts = min_attempts
```

Comment:
```suggestion
        self._min_attempts = max(min_attempts, 1)
```

Reply:
Good point; given that this is specified manually I've decide to treat it as an error to force it to be fixed instead of just allowing (and ignoring) and incorrect value.